git/filter_protocol_test.go

```diff
@@ -0,0 +1,211 @@
+		In: []byte{
+			0x30, 0x30, 0x30, 0x34, // 0004 (hex. length)
+			// No body
+		},
+		Err: "Invalid packet length.",
+	}
+
+	tc.Assert(t)
+}
+
+func TestFilterProtocolDiscardsPacketsWithUnparseableLength(t *testing.T) {
+	tc := &PacketReadTestCase{
+		In: []byte{
+			0xff, 0xff, 0xff, 0xff, // ÿÿÿÿ (invalid hex. length)
+			// No body
+		},
+		Err: "strconv.ParseInt: parsing \"\\xff\\xff\\xff\\xff\": invalid syntax",
+	}
+
+	tc.Assert(t)
+}
+
+func TestFilterProtocolDiscardsPacketsWithLengthZero(t *testing.T) {
+	tc := &PacketReadTestCase{
+		In: []byte{
+			0x30, 0x30, 0x30, 0x30, // 0000 (hex. length)
+			// Empty body
+		},
+	}
+
+	tc.Assert(t)
+}
+
+func TestFilterProtocolReadsTextWithNewline(t *testing.T) {
+	rw := newProtocolRW(bytes.NewReader([]byte{
+		0x30, 0x30, 0x30, 0x39, // 0009 (hex. length)
+		0x61, 0x62, 0x63, 0x64, 0xa,
+		// Empty body
+	}), nil)
+
+	str, err := rw.readPacketText()
+
+	assert.Nil(t, err)
+	assert.Equal(t, "abcd", str)
+}
+
+func TestFilterProtocolReadsTextWithoutNewline(t *testing.T) {
+	rw := newProtocolRW(bytes.NewReader([]byte{
+		0x30, 0x30, 0x30, 0x38, // 0009 (hex. length)
+		0x61, 0x62, 0x63, 0x64,
+	}), nil)
+
+	str, err := rw.readPacketText()
+
+	assert.Nil(t, err)
+	assert.Equal(t, "abcd", str)
+}
+
+func TestFilterProtocolReadsTextWithErr(t *testing.T) {
+	rw := newProtocolRW(bytes.NewReader([]byte{
+		0x30, 0x30, 0x30, 0x34, // 0004 (hex. length)
+		// No body
+	}), nil)
+
+	str, err := rw.readPacketText()
+
+	require.NotNil(t, err)
+	assert.Equal(t, "Invalid packet length.", err.Error())
+	assert.Equal(t, "", str)
+}
+
+func TestFilterProtocolAppendsPacketLists(t *testing.T) {
+	rw := newProtocolRW(bytes.NewReader([]byte{
+		0x30, 0x30, 0x30, 0x38, // 0009 (hex. length)
+		0x61, 0x62, 0x63, 0x64, // "abcd"
+
+		0x30, 0x30, 0x30, 0x38, // 0008 (hex. length)
+		0x65, 0x66, 0x67, 0x68, // "efgh"
+
+		0x30, 0x30, 0x30, 0x30, // 0000 (hex. length)
+	}), nil)
+
+	str, err := rw.readPacketList()
+
+	assert.Nil(t, err)
+	assert.Equal(t, []string{"abcd", "efgh"}, str)
+}
+
+func TestFilterProtocolAppendsPacketListsAndReturnsErrs(t *testing.T) {
+	rw := newProtocolRW(bytes.NewReader([]byte{
+		0x30, 0x30, 0x30, 0x38, // 0009 (hex. length)
+		0x61, 0x62, 0x63, 0x64, // "abcd"
+
+		0x30, 0x30, 0x30, 0x34, // 0004 (hex. length)
+		// No body
+	}), nil)
+
+	str, err := rw.readPacketList()
+
+	require.NotNil(t, err)
+	assert.Equal(t, "Invalid packet length.", err.Error())
+	assert.Empty(t, str)
+}
+
+func TestFilterProtocolWritesPackets(t *testing.T) {
+	var buf bytes.Buffer
+
+	rw := newProtocolRW(nil, &buf)
+	err := rw.writePacket([]byte{
+		0x1, 0x2, 0x3, 0x4,
+	})
+
+	assert.Nil(t, err)
+	assert.Equal(t, []byte{
+		0x30, 0x30, 0x30, 0x38, // 0008 (hex. length)
+		0x1, 0x2, 0x3, 0x4, // payload
+	}, buf.Bytes())
+}
+
+func TestFilterProtocolDoesNotWritePacketsExceedingMaxLength(t *testing.T) {
+	var buf bytes.Buffer
+
+	rw := newProtocolRW(nil, &buf)
+	err := rw.writePacket(make([]byte, MaxPacketLength+1))
```

Comment:
Good call! Added one such test in https://github.com/github/git-lfs/pull/1619/commits/91e54e36f6cca92edf9f13b860fdbac07d987d66.